lib/util/uint128_t.h

```diff
@@ -124,7 +124,11 @@ class uint128_t{
         {
             if (std::is_signed<T>::value) {
                 if (rhs < 0) {
+#ifdef ALLOW_UINT128_FROM_NEGATIVE_NUMBERS
                     UPPER = -1;
+#else
```

Comment:
I'm not actually sure this is wise -- if uint128_t does the check itself, it guarantees the check, whereas the user could perform the cast _without_ checking in the code you pasted here -- but I'll give it a try and see how a variant like this reads.

Reply:
What I am saying is that we're not supposed to use signed arithmetic outside of those helper functions.
Basically, it's much easier to spot a bad type conversion in code reviews (forced because "it won't compile") than an accidental use that will trip only at runtime in (potentially obscure) edge cases.

Reply:
Fair enough. I have a variant that works this way compiling now, just testing.